_posts/2017-07-26-property-based-testing-in-javascript.md

```diff
@@ -0,0 +1,191 @@
+categories: javascript testing
+author_name: Stefan Lau
+author_url : /author/stefanlau
+author_avatar: stefanlau
+show_avatar : true
+read_time : 10
+show_related_posts: false
+---
+
+As developers we are always looking for new ways to ensure that our code works as we would expected. One way to
+achieve this, which is not as commonly used as for example unit tests, is property-based testing. Property-based testing
+is a testing method based on (pseudo-)randomly generated input. It can be used at any level of the testing pyramid to
+validate some properties of algorithms or systems we built and is best used in addition to other tests, which test
+specific instances of inputs.
+
+To create a test case you need to define the following parts:
+
+- **Unit under test**: E.g. a function, endpoint in your API
+- **Input generator function**: A function that will generate pseudo-random input for your test case (based on a seed)
+- **A property predicate**: A predicate that has to hold true for all inputs generated by the generator function.
+  It should include the execution of your unit under test
+
+The test runner will then roughly do the following when executing your test:
+
+1. Ideally it will seed your generator function with a random seed. This is necessary to make the results reproducible
+1. It will feed the next output of your generator function into the property predicate
+1. It will check whether the output of your unit-under-test passes the property predicate
+   - If false: fail test and print the input that lead to non-compliance with the property description
+   - If true: continue with 2. (until some number of iterations is reached)
+
+A simple example for a property based test would be the following property based test for the plus function (for
+simplicities sake I left out the seed part of the generator function). This property-based test tests whether the
+sum of two positive numbers is always positive:
+
+- **Unit under test**: `const add = (x, y) => x + y`
+- **Input generator function**: `const generatePositiveInt = () => Math.round(Math.random() * 100000)`
+- **Property predicate**: `const prop = (in1, in2) => add(in1, in2) >= 0`
+
+The advantage of this type of testing is that your function will be tested with way more input values compared to testing
+with regular unit tests. This allows you to find edge cases where your unit-under-test does not behave as you would expect
+it to. This is quite similar to Fuzzing in the IT-Security community, where the goal is to find security issues or memory
+leaks by providing (invalid) random data.
+
+A disadvantage is that, depending on the unit-under-test, the properties can be quite hard to define. For the simple
+plus example from before it was quite easy, but different algorithms have different properties and they are often very unique
+to the unit-under-test. Another drawback is that property-based testing does not guaranteed to find any existing issues with any
```

Comment:
I'm not quite sure I follow. First of: are you talking about pros and cons of property based tests in contrast to unit tests? Furthermore: 

> Another drawback is that property-based testing does not guaranteed to find any existing issues with any

How do unit tests give you any guarantee that you are finding problems with the cases that you came up with?